### PR TITLE
[filters] Add RBAC condition fuzzer (CEL expressions)

### PR DESCRIPTION
**Pull request by asraa**
This adds a fuzzer for CEL expression matching. These conditions are used in the RBAC filter for complementing the existing principal/permission model.

About a quarter of the execution time is coming from google::api::expr::runtime::RegisterBuiltinFunctions. The test runs locally at about 250 exec/sec.

See: #7716

Risk Level: Low
Testing: Converted unit tests into corpus entries

Re-open of #8752 -- all history the same, merged to up to date master, and pushed all changes. 
/review @htuch

### CODE REVIEW

File: bazel/repository_locations.bzl [hunk 1]
```diff
@@ -94,10 +94,10 @@ REPOSITORY_LOCATIONS = dict(
         urls = ["https://github.com/gabime/spdlog/archive/v1.3.1.tar.gz"],
     ),
     com_github_google_libprotobuf_mutator = dict(
-        sha256 = "f45c3ad82376d891cd0bcaa7165e83efd90e0014b00aebf0cbaf07eb05a1d3f9",
-        strip_prefix = "libprotobuf-mutator-d1fe8a7d8ae18f3d454f055eba5213c291986f21",
-        # 2019-07-10
-        urls = ["https://github.com/google/libprotobuf-mutator/archive/d1fe8a7d8ae18f3d454f055eba5213c291986f21.tar.gz"],
+        sha256 = "54597f640c0ab5e5d783d2f3d3cfe8ad6da999ef1a194d89c2c5ab89a1fd8e13",
+        strip_prefix = "libprotobuf-mutator-dd89da92b59b1714bab6e2a135093948a1cf1c6a",
+        # 2019-10-08
```
Comment (htuch) [inline bazel/repository_locations.bzl]: Why is this needed?

File: test/extensions/filters/common/expr/evaluator_fuzz_test.cc [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+namespace Envoy {
+namespace Extensions {
+namespace Filters {
+namespace Common {
+namespace Expr {
+namespace {
+
+// Set stream_info using information from input.
+// Duplicates most of fromStreamInfo, except is more lightweight without the use of mocks.
```
Comment (htuch) [inline test/extensions/filters/common/expr/evaluator_fuzz_test.cc]: Should this be a general fuzzing util in `test/fuzz/utility.cc` or somewhere like that?

File: test/extensions/filters/common/expr/evaluator_fuzz_test.cc [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+  stream_info.metadata_ = info.dynamic_metadata();
+  const auto start_time =
+      static_cast<uint64_t>(std::numeric_limits<std::chrono::nanoseconds::rep>::max()) <
+              info.start_time()
+          ? 0
+          : info.start_time() / 1000;
+  stream_info.start_time_ = SystemTime(std::chrono::microseconds(start_time));
+  stream_info.setRequestedServerName(info.requested_server_name());
+  stream_info.response_code_ = info.has_response_code() ? info.response_code().value() : 200;
+  auto address = Network::Utility::resolveUrl("tcp://10.0.0.1:443");
```
Comment (htuch) [inline test/extensions/filters/common/expr/evaluator_fuzz_test.cc]: Why not fuzz the address?

File: test/extensions/filters/common/expr/evaluator_fuzz_test.cc [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    TestStreamInfo stream_info = makeStreamInfo(input.stream_info());
+    Http::TestHeaderMapImpl request_headers = Fuzz::fromHeaders(input.request_headers());
+    Http::TestHeaderMapImpl response_headers = Fuzz::fromHeaders(input.response_headers());
+    Http::TestHeaderMapImpl response_trailers = Fuzz::fromHeaders(input.trailers());
+
+    // Evaluate the CEL expression.
+    Protobuf::Arena arena;
+    Expr::evaluate(*expr, nullptr, stream_info, &request_headers, &response_headers,
+                   &response_trailers);
+  } catch (const EnvoyException& e) {
```
Comment (htuch) [inline test/extensions/filters/common/expr/evaluator_fuzz_test.cc]: Does this only happen during proto validation? If so, maybe move `Expr::evaluate` outside the try/catch, to allow the fuzzer to discover any surprise evaluation exceptions.

File: test/extensions/filters/common/expr/evaluator_fuzz_test.cc (hunk 3 above)
Comment (asraa) [reply test/extensions/filters/common/expr/evaluator_fuzz_test.cc]: Done! I decided it might be good to fuzz the way we consume the address anyway.

File: test/extensions/filters/common/expr/evaluator_fuzz_test.cc (hunk 2 above)
Comment (asraa) [reply test/extensions/filters/common/expr/evaluator_fuzz_test.cc]: Done. I added a TODO to make it faster. I originally didn't have it use the existing method because of the mock usage, but that should be fixed across the board.

File: bazel/repository_locations.bzl (hunk 1 above)
Comment (asraa) [reply bazel/repository_locations.bzl]: I'll need to investigate this, but running with the fuzzer in the older version gives me this asan error (which is showing up on OSS-Fuzz)

```
evaluator_fuzz_test_with_libfuzzer: external/com_github_google_libprotobuf_mutator/src/mutator.cc:487: void protobuf_mutator::Mutator::Mutate(google::protobuf::Message *, size_t): Assertion `!keep_initialized_ || message->IsInitialized()' failed.
AddressSanitizer:DEADLYSIGNAL
=================================================================
==140620==ERROR: AddressSanitizer: ABRT on unknown address 0x6fad50002254c (pc 0x7f9cb7bdccfb bp 0x7f9cb7d29bc0 sp 0x7ffe9b1c9fd0 T0)
    #0 0x7f9cb7bdccfa in raise (/lib/x86_64-linux-gnu/libc.so.6+0x36cfa)
    #1 0x7f9cb7bc78ac in abort (/lib/x86_64-linux-gnu/libc.so.6+0x218ac)
    #2 0x7f9cb7bc777e  (/lib/x86_64-linux-gnu/libc.so.6+0x2177e)
    #3 0x7f9cb7bd5541 in __assert_fail (/lib/x86_64-linux-gnu/libc.so.6+0x2f541)
    #4 0xfa98e3d in protobuf_mutator::Mutator::Mutate(google::protobuf::Message*, unsigned long) (/usr/local/google/home/asraa/.cache/bazel/_bazel_asraa/f23f95d44d486dfd95cb67d4d7eb464d/execroot/envoy/bazel-out/k8-fastbuild/bin/test/extensions/filters/common/expr/evaluator_fuzz_test_with_libfuzzer+0xfa98e3d)
```

File: test/extensions/filters/common/expr/evaluator_fuzz_test.cc (hunk 4 above)
Comment (asraa) [reply test/extensions/filters/common/expr/evaluator_fuzz_test.cc]: Ahh... it's an awkward case. The CEL expressions that are fuzzed have some required proto constraints that aren't enforced in proto validations, so they'll get caught in the code by an exception (for example, an `expr_kind` must be set).

File: bazel/repository_locations.bzl (hunk 1 above)
Comment (htuch) [reply bazel/repository_locations.bzl]: Yeah, I think it's fine to bump if it fixes.

File: test/extensions/filters/common/expr/evaluator_fuzz_test.cc (hunk 4 above)
Comment (htuch) [reply test/extensions/filters/common/expr/evaluator_fuzz_test.cc]: Can we add a dedicated `EnvoyException` subclass for this and do a narrower catch around that?

**Comment by kyessenov** [comment]
Thanks! @tswadell FYI.

File: test/extensions/filters/common/expr/evaluator_fuzz_test.cc (hunk 4 above)
Comment (asraa) [reply test/extensions/filters/common/expr/evaluator_fuzz_test.cc]: Done! We throw CelException's for errors thrown in the CEL library
